hawc/apps/animal/exports.py

```diff
@@ -268,6 +268,174 @@ def prepare_df(self, df):
+                )
+            )
+            .filter(
+                stdev__isnull=False,
+            )
+            .values(
+                "endpoint_id",
+                "dose_group_id",
+                "n",
+                "response",
+                "stdev",
+                dose_regime_id=F("endpoint__animal_group__dosing_regime_id"),
+            )
+            .order_by("endpoint_id", "dose_regime_id", "dose_group_id")
+        )
+        return pd.DataFrame(data=qs)
+
+    def get_dichotomous_datasets(self, doses: pd.DataFrame, responses: pd.DataFrame) -> list[dict]:
+        if responses.empty or doses.empty:
+            return []
+
+        df1 = (
+            responses[["endpoint_id", "dose_regime_id", "dose_group_id"]]
+            .drop_duplicates()
+            .set_index(["dose_regime_id", "dose_group_id"])
+        )
+
+        df2 = doses.set_index(["dose_regime_id", "dose_group_id"])
+
+        df3 = (
+            df1.merge(df2, how="inner", left_index=True, right_index=True)
+            .reset_index()
+            .sort_values(["endpoint_id", "dose_units__name", "dose_group_id"])
+            .drop(columns=["dose_regime_id", "dose_group_id"])
+            .groupby(["endpoint_id", "dose_units__name"])
+            .agg(list)
+            .reset_index()
+            .set_index("endpoint_id")
+            .rename(columns={"dose_units__name": "units", "dose": "doses"})
+        )
+
+        df4 = (
+            responses.set_index("endpoint_id")
+            .loc[df3.index.values]
+            .drop(columns=["dose_regime_id", "dose_group_id"])
+            .groupby(["endpoint_id"])
+            .agg(list)
+            .assign(dtype="D")
+            .reset_index()
+        )
+        results = df4.to_dict(orient="records")
+        for result in results:
+            d = df3.loc[result["endpoint_id"]]
+            if isinstance(d, pd.Series):
+                d = d.to_frame().T
+            result["doses"] = {d["units"]: d["doses"] for d in d.to_dict(orient="records")}
+
+        return results
+
+    def get_continuous_datasets(self, doses, responses) -> list[dict]:
+        if responses.empty or doses.empty:
+            return []
+
+        df1 = (
+            responses[["endpoint_id", "dose_regime_id", "dose_group_id"]]
+            .drop_duplicates()
+            .set_index(["dose_regime_id", "dose_group_id"])
+        )
+
+        df2 = doses.set_index(["dose_regime_id", "dose_group_id"])
+
+        df3 = (
+            df1.merge(df2, how="inner", left_index=True, right_index=True)
+            .reset_index()
+            .sort_values(["endpoint_id", "dose_units__name", "dose_group_id"])
+            .drop(columns=["dose_regime_id", "dose_group_id"])
+            .groupby(["endpoint_id", "dose_units__name"])
+            .agg(list)
+            .reset_index()
+            .set_index("endpoint_id")
+            .rename(columns={"dose_units__name": "units", "dose": "doses"})
+        )
+
+        df4 = (
```

Comment:
Are there any names for these intermediate dataframes that would make sense instead of numbering them df1, df2, etc.? That would make it a little easier to follow